src/goto-symex/goto_symex.cpp

```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+  const array_exprt &new_char_array)
+{
+  // assign length of string
+  symex_assign.assign_symbol(length, expr_skeletont{}, new_length, {});
+
+  const std::string aux_symbol_name =
+    get_alnum_string(new_char_array) + "_constant_char_array";
+
+  const bool string_constant_exists =
+    state.symbol_table.has_symbol(aux_symbol_name);
+
+  const symbolt &aux_symbol =
+    string_constant_exists
+      ? state.symbol_table.lookup_ref(aux_symbol_name)
+      : get_new_string_data_symbol(
+          state, symex_assign, aux_symbol_name, char_array, new_char_array);
+
+  INVARIANT(
+    aux_symbol.value == new_char_array,
+    "symbol shall have value derived from char array content");
+
+  const address_of_exprt string_data(
+    index_exprt(aux_symbol.symbol_expr(), from_integer(0, index_type())));
+
+  symex_assign.assign_symbol(char_array, expr_skeletont{}, string_data, {});
+
+  if(!string_constant_exists)
+  {
+    associate_array_to_pointer(
+      state, symex_assign, new_char_array, string_data);
+  }
+}
+
+const symbolt &goto_symext::get_new_string_data_symbol(
+  statet &state,
+  symex_assignt &symex_assign,
+  const std::string &aux_symbol_name,
+  const ssa_exprt &char_array,
+  const array_exprt &new_char_array)
+{
+  array_typet new_char_array_type = new_char_array.type();
+  new_char_array_type.set(ID_C_constant, true);
+
+  symbolt &new_aux_symbol = get_fresh_aux_symbol(
+    new_char_array_type,
+    "",
+    aux_symbol_name,
+    source_locationt(),
+    ns.lookup(to_symbol_expr(char_array.get_original_expr())).mode,
+    ns,
+    state.symbol_table);
+
+  CHECK_RETURN(new_aux_symbol.is_state_var);
+  CHECK_RETURN(new_aux_symbol.is_lvalue);
+
+  new_aux_symbol.is_static_lifetime = true;
+  new_aux_symbol.is_file_local = false;
+  new_aux_symbol.is_thread_local = false;
+
+  new_aux_symbol.value = new_char_array;
+
+  const exprt arr = state.rename(new_aux_symbol.symbol_expr(), ns).get();
+
+  symex_assign.assign_symbol(
+    to_ssa_expr(arr).get_l1_object(), expr_skeletont{}, new_char_array, {});
+
+  return new_aux_symbol;
+}
+
+void goto_symext::associate_array_to_pointer(
+  statet &state,
+  symex_assignt &symex_assign,
+  const array_exprt &new_char_array,
+  const address_of_exprt &string_data)
+{
+  const symbolt &function_symbol =
+    ns.lookup(ID_cprover_associate_array_to_pointer_func);
+
+  const function_application_exprt array_to_pointer_app{
+    function_symbol.symbol_expr(), {new_char_array, string_data}};
+
+  const symbolt &return_symbol = get_fresh_aux_symbol(
+    to_mathematical_function_type(function_symbol.type).codomain(),
+    "",
+    "return_value",
+    source_locationt(),
+    function_symbol.mode,
+    ns,
+    state.symbol_table);
+
+  const ssa_exprt ssa_expr(return_symbol.symbol_expr());
+
+  symex_assign.assign_symbol(
+    ssa_expr, expr_skeletont{}, array_to_pointer_app, {});
+}
+
+optionalt<std::reference_wrapper<const array_exprt>>
+goto_symext::try_evaluate_constant_string(
```

Comment:
🐙

Reply:
I prefer those methods to live in `goto_symext` as opposed to being static. The plan is to eventually create a `goto_symex_constant_propagationt` member of `goto_symext` which handles the constant propagation.